milabench/executors.py

```diff
@@ -0,0 +1,465 @@
+
+    def _argv(self, **kwargs) -> List:
+        del kwargs
+        return []
+
+
+# Leafs
+class CmdExecutor(Executor):
+    def __init__(
+            self,
+            pack:"pack.BasePackage",
+            *cmd_argv,
+            **kwargs
+    ) -> None:
+        if isinstance(pack, Executor):
+            raise ValueError(f"{self.__class__} does not accept nested"
+                             f" {Executor.__class__}")
+        super().__init__(
+            pack,
+            **kwargs
+        )
+        self.bin_argv = cmd_argv
+
+    def _argv(self, **kwargs) -> List:
+        del kwargs
+        return [*self.bin_argv]
+
+
+class PackExecutor(CmdExecutor):
+    def __init__(
+            self,
+            pack:"pack.BasePackage",
+            *script_argv,
+            **kwargs
+    ) -> None:
+        script = script_argv[:1]
+        if script and XPath(script[0]).exists():
+            script = script[0]
+            script_argv = script_argv[1:]
+        else:
+            script = None
+
+        super().__init__(
+            pack,
+            *script_argv,
+            **kwargs
+        )
+
+        self.script = script
+
+    def _argv(self, **kwargs) -> List:
+        main = self.script or self.pack.main_script
+        if not XPath(main).is_absolute():
+            main = self.pack.dirs.code / main   # Could this lead any unexpected
+                                                # path during exec?
+
+        if not main.exists():
+            raise FileNotFoundError(
+                f"Cannot run main script because it does not exist: {main}"
+            )
+
+        if main.is_dir():
+            main = ["-m", str(self.pack.main_script)]
+        else:
+            main = [str(main)]
+        return main + super()._argv(**kwargs)
+
+
+class VoidExecutor(Executor):
+    _VoidPack=None
+    _VOIDPACK=None
+
+    def __new__(cls, *args, **kwargs):
+        if cls._VOIDPACK is None:
+            class _VoidPack(pack.BasePackage):
+                def __init__(self, *args, **kwargs):
+                    del args, kwargs
+                    pass
+
+                def execute(self, *args, **kwargs):
+                    del args, kwargs
+                    pass
+            cls._VoidPack = _VoidPack
+            cls._VOIDPACK = cls._VoidPack()
+
+        return super().__new__(*args, **kwargs)
+
+    def __init__(self, *args, **kwargs) -> None:
+        del args, kwargs
+        super().__init__(VoidExecutor._VOIDPACK)
+
+
+# Branches or Roots
+class DockerRunExecutor(Executor):
+    def __init__(
+            self,
+            executor:Executor,
+            image:str,
+            *docker_argv,
+            **kwargs
+    ) -> None:
+        super().__init__(
+            executor,
+            **kwargs
+        )
+        self.image = image
+        self.docker_argv = docker_argv
+
+    def _argv(self, **kwargs) -> List:
+        del kwargs
+        # TODO: If in docker: no-op
+        # if :
```

Comment:
```suggestion
        if os.path.exists('/.dockerenv'),:
```

Reply:
Sorry, the code is all over the place but in #134 because of https://github.com/moby/moby/issues/18355#issuecomment-220484748 I've chosen to simoly use the env var :
```
        if self.image is None or os.environ.get("MILABENCH_DOCKER", None):
            # No-op when there's no docker image to run or inside a docker
            # container
            return []
```